runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -6710,6 +6795,123 @@ int amo_cmd_2_nic_op(fork_amo_cmd_t cmd, int fetching)
   return nic_cmd;
 }
 
+#if HAVE_GNI_FMA_CHAIN_TRANSACTIONS
+#define MAX_BUFF_AMO_THREADS 1024
```

Comment:
This aspect of the design troubles me.  These are statically sized and have the potential to get used up during execution.  (The likely way that would happen is probably not long-lived tasking layer pthreads because we won't ever create 1k of those even with tasks=fifo, but rather other pthreads from other programming models or the user, coming in here from non-Chapel code calling Chapel code in a multi-language program.  Once an entry is used here it's never freed, so a program that creates a pthread which calls Chapel code, ends up grabbing an entry here, and then terminates will leave the entry dangling and unusable.)  At the very least we should emit a warning when they do get used up, since we're going to fall off a performance cliff for super-relaxed AMOs at that point.  As an alternative design that doesn't have static sizing, how about defining a thread-private struct containing all the meta-information for each thread's buffered AMOs, and then just chain those together on a (doubly) linked list, under protection of a lock?  Adding the thread-private struct to the list could be done under an `inited` flag in `do_nic_amo_nf_buff()` where you do initialization now.  You'd also need to create a pthread cleanup function (see `othread_cleanup_push()`) to remove the thread's struct from the chain if/when it terminated.

Reply:
Offline we agreed to emit a warning for now, and we can address this later -- practically speaking this is unlikely to bite us in the near future (at least for Chapel-only programs.)